snuba/replacer.py

```diff
@@ -16,13 +21,138 @@
+        self.__table = table
+        self.__metrics = metrics
+        self.__runner = runner
+
+    def execute(self, replacement: Replacement, records_count: int) -> int:
+        query = replacement.get_insert_query(self.__table)
+        if query is None:
+            return 0
+        self.__runner(self.__connection, query, records_count, self.__metrics)
+        return records_count
+
+
+class LocalTableExecutor(InsertExecutor):
+    """
+    Executes a replacement query on each individual shard in parallel.
+
+    It implements some basic retry logic by trying a different replica
+    if the first attempt fails.
+    It also falls back on a DistributedExecutor if everything fails.
+    """
+
+    def __init__(
+        self,
+        runner: RunQuery,
+        thread_pool: ThreadPoolExecutor,
+        main_connections: Mapping[int, Sequence[ClickhousePool]],
+        main_table: str,
+        backup_executor: InsertExecutor,
+        metrics: MetricsBackend,
+    ) -> None:
+        self.__thread_pool = thread_pool
+        self.__main_connections = main_connections
+        self.__main_table = main_table
+        self.__backup_executor = backup_executor
+        self.__metrics = metrics
+        self.__runner = runner
+
+    def __run_multiple_replicas(
+        self,
+        connections: Sequence[ClickhousePool],
+        query: str,
+        records_count: int,
+        metrics: MetricsBackend,
+    ) -> None:
+        """
+        Makes multiple attempts to run the query.
+        One per connection provided.
+        """
+        for attempt in range(len(connections)):
+            try:
+                self.__runner(connections[attempt], query, records_count, metrics)
+                return
+            except Exception as e:
+                if attempt == len(connections) - 1:
+                    raise
+                logger.warning(
+                    "Replacement processing failed on the main connection", exc_info=e,
+                )
+
+    def execute(self, replacement: Replacement, records_count: int) -> int:
+        try:
+            query = replacement.get_insert_query(self.__main_table)
+            if query is None:
+                return 0
+            result_futures = []
+            for connections in self.__main_connections.values():
+                result_futures.append(
+                    self.__thread_pool.submit(
+                        partial(
+                            self.__run_multiple_replicas,
+                            connections=connections,
+                            query=query,
+                            records_count=records_count,
+                            metrics=self.__metrics,
+                        )
+                    )
+                )
+            for result in as_completed(result_futures):
+                e = result.exception()
+                if e is not None:
+                    raise e
+            return records_count
```

Comment:
This idea of running a query once per shard and retrying on a number of replicas may actually be useful outside of replacements too. Migrations are an example of where it could used. But they'll probably be significant differences over there so probably not worth generalising for that.

Reply:
I did not generalise that here because the table executor knows too much context about replacements. 
But I agree we could and should. Though that would likely add around 500 lines.